{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f378d645-9130-4fdc-968c-81b76e3b5acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "79e5b190-41a5-45c0-ab48-9dcb719cd834",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "634869fa-82e9-459d-a20f-6aa2784e9c58",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/b_/drg8gqvj38n0vbrfqvqz3k1c0000gn/T/ipykernel_52831/622971896.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('/Users/kinam/git/obds_sep2022/Python_W2/Homo_sapiens.GRCh38.107.gtf', sep='\\t', header = None, comment = \"#\") # Read a tsv file\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('/Users/kinam/git/obds_sep2022/Python_W2/Homo_sapiens.GRCh38.107.gtf', sep='\\t', header = None, comment = \"#\") # Read a tsv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "831267c6-3f79-4c98-a58f-b4fbc1e74897",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0               1            2        3        4  5  6  7  \\\n",
       "0  1  ensembl_havana         gene  1471765  1497848  .  +  .   \n",
       "1  1  ensembl_havana   transcript  1471765  1497848  .  +  .   \n",
       "2  1  ensembl_havana         exon  1471765  1472089  .  +  .   \n",
       "3  1  ensembl_havana          CDS  1471885  1472089  .  +  0   \n",
       "4  1  ensembl_havana  start_codon  1471885  1471887  .  +  0   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b95a1a71-30b7-4b2f-946c-a15f90df7fa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'group']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2daf3103-d7da-4000-b0e9-1a6d5d9fd858",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source      feature    start      end score strand frame  \\\n",
       "0       1  ensembl_havana         gene  1471765  1497848     .      +     .   \n",
       "1       1  ensembl_havana   transcript  1471765  1497848     .      +     .   \n",
       "2       1  ensembl_havana         exon  1471765  1472089     .      +     .   \n",
       "3       1  ensembl_havana          CDS  1471885  1472089     .      +     0   \n",
       "4       1  ensembl_havana  start_codon  1471885  1471887     .      +     0   \n",
       "\n",
       "                                               group  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "30f896b8-a9d9-451c-9b80-e0456cc84b54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3371244, 9)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dbc3eddc-78d5-4a0f-be57-cbc841bbe110",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df['source'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3873ed3e-e584-4270-9582-aa3a4e389afe",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_df = df.query(\"source == 'ensembl_havana'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c9b830e2-b94a-45ff-bc84-df2044fd381f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(714608, 9)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cc8eed7f-f48d-4fb1-b673-73cddc698114",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source      feature    start      end score strand frame  \\\n",
       "0       1  ensembl_havana         gene  1471765  1497848     .      +     .   \n",
       "1       1  ensembl_havana   transcript  1471765  1497848     .      +     .   \n",
       "2       1  ensembl_havana         exon  1471765  1472089     .      +     .   \n",
       "3       1  ensembl_havana          CDS  1471885  1472089     .      +     0   \n",
       "4       1  ensembl_havana  start_codon  1471885  1471887     .      +     0   \n",
       "\n",
       "                                               group  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b9c33a6c-cfc4-4dfb-a420-51a8d0062238",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>871196</td>\n",
       "      <td>871196</td>\n",
       "      <td>871196</td>\n",
       "      <td>871196</td>\n",
       "      <td>871196</td>\n",
       "      <td>871196</td>\n",
       "      <td>871196</td>\n",
       "      <td>871196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Selenocysteine</th>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>1624585</td>\n",
       "      <td>1624585</td>\n",
       "      <td>1624585</td>\n",
       "      <td>1624585</td>\n",
       "      <td>1624585</td>\n",
       "      <td>1624585</td>\n",
       "      <td>1624585</td>\n",
       "      <td>1624585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_utr</th>\n",
       "      <td>171504</td>\n",
       "      <td>171504</td>\n",
       "      <td>171504</td>\n",
       "      <td>171504</td>\n",
       "      <td>171504</td>\n",
       "      <td>171504</td>\n",
       "      <td>171504</td>\n",
       "      <td>171504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>61860</td>\n",
       "      <td>61860</td>\n",
       "      <td>61860</td>\n",
       "      <td>61860</td>\n",
       "      <td>61860</td>\n",
       "      <td>61860</td>\n",
       "      <td>61860</td>\n",
       "      <td>61860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>96934</td>\n",
       "      <td>96934</td>\n",
       "      <td>96934</td>\n",
       "      <td>96934</td>\n",
       "      <td>96934</td>\n",
       "      <td>96934</td>\n",
       "      <td>96934</td>\n",
       "      <td>96934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>90724</td>\n",
       "      <td>90724</td>\n",
       "      <td>90724</td>\n",
       "      <td>90724</td>\n",
       "      <td>90724</td>\n",
       "      <td>90724</td>\n",
       "      <td>90724</td>\n",
       "      <td>90724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_utr</th>\n",
       "      <td>203201</td>\n",
       "      <td>203201</td>\n",
       "      <td>203201</td>\n",
       "      <td>203201</td>\n",
       "      <td>203201</td>\n",
       "      <td>203201</td>\n",
       "      <td>203201</td>\n",
       "      <td>203201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>251121</td>\n",
       "      <td>251121</td>\n",
       "      <td>251121</td>\n",
       "      <td>251121</td>\n",
       "      <td>251121</td>\n",
       "      <td>251121</td>\n",
       "      <td>251121</td>\n",
       "      <td>251121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seqname   source    start      end    score   strand  \\\n",
       "feature                                                                 \n",
       "CDS               871196   871196   871196   871196   871196   871196   \n",
       "Selenocysteine       119      119      119      119      119      119   \n",
       "exon             1624585  1624585  1624585  1624585  1624585  1624585   \n",
       "five_prime_utr    171504   171504   171504   171504   171504   171504   \n",
       "gene               61860    61860    61860    61860    61860    61860   \n",
       "start_codon        96934    96934    96934    96934    96934    96934   \n",
       "stop_codon         90724    90724    90724    90724    90724    90724   \n",
       "three_prime_utr   203201   203201   203201   203201   203201   203201   \n",
       "transcript        251121   251121   251121   251121   251121   251121   \n",
       "\n",
       "                   frame    group  \n",
       "feature                            \n",
       "CDS               871196   871196  \n",
       "Selenocysteine       119      119  \n",
       "exon             1624585  1624585  \n",
       "five_prime_utr    171504   171504  \n",
       "gene               61860    61860  \n",
       "start_codon        96934    96934  \n",
       "stop_codon         90724    90724  \n",
       "three_prime_utr   203201   203201  \n",
       "transcript        251121   251121  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(\"feature\").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "ee64cffa-94a9-452a-bd21-98513038f07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = df.query(\"feature == 'gene'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "1d238808-ba7f-42f1-958f-13862ada0b1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(61860, 9)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "c3683f84-2f3e-44fc-8112-1d599c5db7f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2212523</td>\n",
       "      <td>2220738</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234396\"; gene_version \"3\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225972\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8786211</td>\n",
       "      <td>8786913</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000224315\"; gene_version \"2\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>634376</td>\n",
       "      <td>634922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198744\"; gene_version \"5\"; g...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname          source feature    start      end score strand frame  \\\n",
       "0         1  ensembl_havana    gene  1471765  1497848     .      +     .   \n",
       "107       1          havana    gene  2212523  2220738     .      +     .   \n",
       "111       1          havana    gene   629062   629433     .      +     .   \n",
       "114       1          havana    gene  8786211  8786913     .      -     .   \n",
       "117       1          havana    gene   634376   634922     .      +     .   \n",
       "\n",
       "                                                 group  \n",
       "0    gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "107  gene_id \"ENSG00000234396\"; gene_version \"3\"; g...  \n",
       "111  gene_id \"ENSG00000225972\"; gene_version \"1\"; g...  \n",
       "114  gene_id \"ENSG00000224315\"; gene_version \"2\"; g...  \n",
       "117  gene_id \"ENSG00000198744\"; gene_version \"5\"; g...  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0ff85dc2-ac0d-4790-b4a8-6de5f70991bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/b_/drg8gqvj38n0vbrfqvqz3k1c0000gn/T/ipykernel_52831/3297691938.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genes['gene_length'] = genes.iloc[:,4] - genes.iloc[:,3]\n"
     ]
    }
   ],
   "source": [
    "# genes['gene_length'] = genes.iloc[:,4] - genes.iloc[:,3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "a4b43fdd-9548-49ac-9034-df4f9af91ba8",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = genes.assign(gene_length = genes[\"end\"] - genes[\"start\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "dc7e46e1-e26e-4ad7-a834-1e2ff8e10716",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(61860, 10)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "50d9bded-deb8-45fb-9922-c804cfd2f5e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2212523</td>\n",
       "      <td>2220738</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234396\"; gene_version \"3\"; g...</td>\n",
       "      <td>8215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225972\"; gene_version \"1\"; g...</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8786211</td>\n",
       "      <td>8786913</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000224315\"; gene_version \"2\"; g...</td>\n",
       "      <td>702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>634376</td>\n",
       "      <td>634922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198744\"; gene_version \"5\"; g...</td>\n",
       "      <td>546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname          source feature    start      end score strand frame  \\\n",
       "0         1  ensembl_havana    gene  1471765  1497848     .      +     .   \n",
       "107       1          havana    gene  2212523  2220738     .      +     .   \n",
       "111       1          havana    gene   629062   629433     .      +     .   \n",
       "114       1          havana    gene  8786211  8786913     .      -     .   \n",
       "117       1          havana    gene   634376   634922     .      +     .   \n",
       "\n",
       "                                                 group  gene_length  \n",
       "0    gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "107  gene_id \"ENSG00000234396\"; gene_version \"3\"; g...         8215  \n",
       "111  gene_id \"ENSG00000225972\"; gene_version \"1\"; g...          371  \n",
       "114  gene_id \"ENSG00000224315\"; gene_version \"2\"; g...          702  \n",
       "117  gene_id \"ENSG00000198744\"; gene_version \"5\"; g...          546  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "1aaa4842-b22c-45d0-a627-6dfd705b443b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_length = genes[\"gene_length\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f1f557c4-3f68-404e-b565-b836846b1465",
   "metadata": {},
   "outputs": [],
   "source": [
    "median_length = genes[\"gene_length\"].median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "e46ed1a2-51c2-4853-8608-4f7256128e89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32626.786744261233"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ff1351bf-91b0-4a65-9e5a-56e4e2b7ac8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3553.5"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "median_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "023e93b8-9a05-48b3-be5a-ff56e8a2fe50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2212523</td>\n",
       "      <td>2220738</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234396\"; gene_version \"3\"; g...</td>\n",
       "      <td>8215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225972\"; gene_version \"1\"; g...</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8786211</td>\n",
       "      <td>8786913</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000224315\"; gene_version \"2\"; g...</td>\n",
       "      <td>702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>634376</td>\n",
       "      <td>634922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198744\"; gene_version \"5\"; g...</td>\n",
       "      <td>546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname          source feature    start      end score strand frame  \\\n",
       "0         1  ensembl_havana    gene  1471765  1497848     .      +     .   \n",
       "107       1          havana    gene  2212523  2220738     .      +     .   \n",
       "111       1          havana    gene   629062   629433     .      +     .   \n",
       "114       1          havana    gene  8786211  8786913     .      -     .   \n",
       "117       1          havana    gene   634376   634922     .      +     .   \n",
       "\n",
       "                                                 group  gene_length  \n",
       "0    gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "107  gene_id \"ENSG00000234396\"; gene_version \"3\"; g...         8215  \n",
       "111  gene_id \"ENSG00000225972\"; gene_version \"1\"; g...          371  \n",
       "114  gene_id \"ENSG00000224315\"; gene_version \"2\"; g...          702  \n",
       "117  gene_id \"ENSG00000198744\"; gene_version \"5\"; g...          546  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "ab74cb2d-ed50-4f51-819c-77bb0237cd85",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_sorted = genes.sort_values(by = ['gene_length'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "36e9ee31-6fa3-467c-89e1-c7d9b9420524",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2341061</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22438547</td>\n",
       "      <td>22438554</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223997\"; gene_version \"1\"; g...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341642</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22439007</td>\n",
       "      <td>22439015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237235\"; gene_version \"2\"; g...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418605</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105865551</td>\n",
       "      <td>105865561</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000236597\"; gene_version \"1\"; g...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312465</th>\n",
       "      <td>7</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>142786213</td>\n",
       "      <td>142786224</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000282431\"; gene_version \"1\"; g...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341080</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22449113</td>\n",
       "      <td>22449125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228985\"; gene_version \"1\"; g...</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  source feature      start        end score strand frame  \\\n",
       "2341061      14  havana    gene   22438547   22438554     .      +     .   \n",
       "2341642      14  havana    gene   22439007   22439015     .      +     .   \n",
       "2418605      14  havana    gene  105865551  105865561     .      -     .   \n",
       "1312465       7  havana    gene  142786213  142786224     .      +     .   \n",
       "2341080      14  havana    gene   22449113   22449125     .      +     .   \n",
       "\n",
       "                                                     group  gene_length  \n",
       "2341061  gene_id \"ENSG00000223997\"; gene_version \"1\"; g...            7  \n",
       "2341642  gene_id \"ENSG00000237235\"; gene_version \"2\"; g...            8  \n",
       "2418605  gene_id \"ENSG00000236597\"; gene_version \"1\"; g...           10  \n",
       "1312465  gene_id \"ENSG00000282431\"; gene_version \"1\"; g...           11  \n",
       "2341080  gene_id \"ENSG00000228985\"; gene_version \"1\"; g...           12  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "8ddb2046-06b4-4933-882d-e82fc9f55bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2341061</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22438547</td>\n",
       "      <td>22438554</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223997\"; gene_version \"1\"; g...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341642</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22439007</td>\n",
       "      <td>22439015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237235\"; gene_version \"2\"; g...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418605</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105865551</td>\n",
       "      <td>105865561</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000236597\"; gene_version \"1\"; g...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312465</th>\n",
       "      <td>7</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>142786213</td>\n",
       "      <td>142786224</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000282431\"; gene_version \"1\"; g...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341080</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22449113</td>\n",
       "      <td>22449125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228985\"; gene_version \"1\"; g...</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418657</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105894508</td>\n",
       "      <td>105894523</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227800\"; gene_version \"1\"; g...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418681</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105903616</td>\n",
       "      <td>105903631</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000232543\"; gene_version \"2\"; g...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418713</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105913222</td>\n",
       "      <td>105913237</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000233655\"; gene_version \"1\"; g...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418701</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105909907</td>\n",
       "      <td>105909923</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237197\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2462756</th>\n",
       "      <td>15</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>20011153</td>\n",
       "      <td>20011169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000271336\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418669</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105900638</td>\n",
       "      <td>105900654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227108\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2467826</th>\n",
       "      <td>15</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>21017800</td>\n",
       "      <td>21017816</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000270185\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418725</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105919502</td>\n",
       "      <td>105919518</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000236170\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418645</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105891191</td>\n",
       "      <td>105891207</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237020\"; gene_version \"1\"; g...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418616</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105881539</td>\n",
       "      <td>105881556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225825\"; gene_version \"1\"; g...</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418705</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105910410</td>\n",
       "      <td>105910427</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228131\"; gene_version \"1\"; g...</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2467325</th>\n",
       "      <td>15</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>21011451</td>\n",
       "      <td>21011469</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000270451\"; gene_version \"1\"; g...</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2489695</th>\n",
       "      <td>15</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>20004797</td>\n",
       "      <td>20004815</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000271317\"; gene_version \"1\"; g...</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418624</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105884870</td>\n",
       "      <td>105884888</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227196\"; gene_version \"1\"; g...</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418653</th>\n",
       "      <td>14</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>105893542</td>\n",
       "      <td>105893561</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000211915\"; gene_version \"1\"; g...</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  source feature      start        end score strand frame  \\\n",
       "2341061      14  havana    gene   22438547   22438554     .      +     .   \n",
       "2341642      14  havana    gene   22439007   22439015     .      +     .   \n",
       "2418605      14  havana    gene  105865551  105865561     .      -     .   \n",
       "1312465       7  havana    gene  142786213  142786224     .      +     .   \n",
       "2341080      14  havana    gene   22449113   22449125     .      +     .   \n",
       "2418657      14  havana    gene  105894508  105894523     .      -     .   \n",
       "2418681      14  havana    gene  105903616  105903631     .      -     .   \n",
       "2418713      14  havana    gene  105913222  105913237     .      -     .   \n",
       "2418701      14  havana    gene  105909907  105909923     .      -     .   \n",
       "2462756      15  havana    gene   20011153   20011169     .      -     .   \n",
       "2418669      14  havana    gene  105900638  105900654     .      -     .   \n",
       "2467826      15  havana    gene   21017800   21017816     .      -     .   \n",
       "2418725      14  havana    gene  105919502  105919518     .      -     .   \n",
       "2418645      14  havana    gene  105891191  105891207     .      -     .   \n",
       "2418616      14  havana    gene  105881539  105881556     .      -     .   \n",
       "2418705      14  havana    gene  105910410  105910427     .      -     .   \n",
       "2467325      15  havana    gene   21011451   21011469     .      -     .   \n",
       "2489695      15  havana    gene   20004797   20004815     .      -     .   \n",
       "2418624      14  havana    gene  105884870  105884888     .      -     .   \n",
       "2418653      14  havana    gene  105893542  105893561     .      -     .   \n",
       "\n",
       "                                                     group  gene_length  \n",
       "2341061  gene_id \"ENSG00000223997\"; gene_version \"1\"; g...            7  \n",
       "2341642  gene_id \"ENSG00000237235\"; gene_version \"2\"; g...            8  \n",
       "2418605  gene_id \"ENSG00000236597\"; gene_version \"1\"; g...           10  \n",
       "1312465  gene_id \"ENSG00000282431\"; gene_version \"1\"; g...           11  \n",
       "2341080  gene_id \"ENSG00000228985\"; gene_version \"1\"; g...           12  \n",
       "2418657  gene_id \"ENSG00000227800\"; gene_version \"1\"; g...           15  \n",
       "2418681  gene_id \"ENSG00000232543\"; gene_version \"2\"; g...           15  \n",
       "2418713  gene_id \"ENSG00000233655\"; gene_version \"1\"; g...           15  \n",
       "2418701  gene_id \"ENSG00000237197\"; gene_version \"1\"; g...           16  \n",
       "2462756  gene_id \"ENSG00000271336\"; gene_version \"1\"; g...           16  \n",
       "2418669  gene_id \"ENSG00000227108\"; gene_version \"1\"; g...           16  \n",
       "2467826  gene_id \"ENSG00000270185\"; gene_version \"1\"; g...           16  \n",
       "2418725  gene_id \"ENSG00000236170\"; gene_version \"1\"; g...           16  \n",
       "2418645  gene_id \"ENSG00000237020\"; gene_version \"1\"; g...           16  \n",
       "2418616  gene_id \"ENSG00000225825\"; gene_version \"1\"; g...           17  \n",
       "2418705  gene_id \"ENSG00000228131\"; gene_version \"1\"; g...           17  \n",
       "2467325  gene_id \"ENSG00000270451\"; gene_version \"1\"; g...           18  \n",
       "2489695  gene_id \"ENSG00000271317\"; gene_version \"1\"; g...           18  \n",
       "2418624  gene_id \"ENSG00000227196\"; gene_version \"1\"; g...           18  \n",
       "2418653  gene_id \"ENSG00000211915\"; gene_version \"1\"; g...           19  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sorted.iloc[0:20,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "eb7c7f08-0aa8-4c4d-837d-42a5221e9cc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_sorted = genes.sort_values(by = ['gene_length'], ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "1a64bc53-afb8-4e0e-a396-8bd8800b5667",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>5239802</td>\n",
       "      <td>7713340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000078328\"; gene_version \"23\"; ...</td>\n",
       "      <td>2473538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>146116002</td>\n",
       "      <td>148420998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000174469\"; gene_version \"23\"; ...</td>\n",
       "      <td>2304996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8314246</td>\n",
       "      <td>10613002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000153707\"; gene_version \"19\"; ...</td>\n",
       "      <td>2298756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198947\"; gene_version \"18\"; ...</td>\n",
       "      <td>2241932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>83455012</td>\n",
       "      <td>85627922</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150672\"; gene_version \"18\"; ...</td>\n",
       "      <td>2172910</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2679474      16  ensembl_havana    gene    5239802    7713340     .      +   \n",
       "1297140       7  ensembl_havana    gene  146116002  148420998     .      +   \n",
       "1733179       9  ensembl_havana    gene    8314246   10613002     .      -   \n",
       "1514121       X  ensembl_havana    gene   31097677   33339609     .      -   \n",
       "1850816      11  ensembl_havana    gene   83455012   85627922     .      -   \n",
       "\n",
       "        frame                                              group  gene_length  \n",
       "2679474     .  gene_id \"ENSG00000078328\"; gene_version \"23\"; ...      2473538  \n",
       "1297140     .  gene_id \"ENSG00000174469\"; gene_version \"23\"; ...      2304996  \n",
       "1733179     .  gene_id \"ENSG00000153707\"; gene_version \"19\"; ...      2298756  \n",
       "1514121     .  gene_id \"ENSG00000198947\"; gene_version \"18\"; ...      2241932  \n",
       "1850816     .  gene_id \"ENSG00000150672\"; gene_version \"18\"; ...      2172910  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "c795d492-1dbd-40b1-bc27-160eab6d061f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>5239802</td>\n",
       "      <td>7713340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000078328\"; gene_version \"23\"; ...</td>\n",
       "      <td>2473538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>146116002</td>\n",
       "      <td>148420998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000174469\"; gene_version \"23\"; ...</td>\n",
       "      <td>2304996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8314246</td>\n",
       "      <td>10613002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000153707\"; gene_version \"19\"; ...</td>\n",
       "      <td>2298756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198947\"; gene_version \"18\"; ...</td>\n",
       "      <td>2241932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>83455012</td>\n",
       "      <td>85627922</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150672\"; gene_version \"18\"; ...</td>\n",
       "      <td>2172910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1601567</th>\n",
       "      <td>8</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2935353</td>\n",
       "      <td>4994972</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183117\"; gene_version \"20\"; ...</td>\n",
       "      <td>2059619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030998</th>\n",
       "      <td>20</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>13995369</td>\n",
       "      <td>16053197</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000172264\"; gene_version \"18\"; ...</td>\n",
       "      <td>2057828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104549</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>63719980</td>\n",
       "      <td>65707226</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188107\"; gene_version \"15\"; ...</td>\n",
       "      <td>1987246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338522</th>\n",
       "      <td>2</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>140231423</td>\n",
       "      <td>142131016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000168702\"; gene_version \"18\"; ...</td>\n",
       "      <td>1899593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041854</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65912457</td>\n",
       "      <td>67763637</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183230\"; gene_version \"18\"; ...</td>\n",
       "      <td>1851180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2679474      16  ensembl_havana    gene    5239802    7713340     .      +   \n",
       "1297140       7  ensembl_havana    gene  146116002  148420998     .      +   \n",
       "1733179       9  ensembl_havana    gene    8314246   10613002     .      -   \n",
       "1514121       X  ensembl_havana    gene   31097677   33339609     .      -   \n",
       "1850816      11  ensembl_havana    gene   83455012   85627922     .      -   \n",
       "1601567       8  ensembl_havana    gene    2935353    4994972     .      -   \n",
       "3030998      20  ensembl_havana    gene   13995369   16053197     .      +   \n",
       "1104549       6  ensembl_havana    gene   63719980   65707226     .      -   \n",
       "338522        2          havana    gene  140231423  142131016     .      -   \n",
       "2041854      10  ensembl_havana    gene   65912457   67763637     .      -   \n",
       "\n",
       "        frame                                              group  gene_length  \n",
       "2679474     .  gene_id \"ENSG00000078328\"; gene_version \"23\"; ...      2473538  \n",
       "1297140     .  gene_id \"ENSG00000174469\"; gene_version \"23\"; ...      2304996  \n",
       "1733179     .  gene_id \"ENSG00000153707\"; gene_version \"19\"; ...      2298756  \n",
       "1514121     .  gene_id \"ENSG00000198947\"; gene_version \"18\"; ...      2241932  \n",
       "1850816     .  gene_id \"ENSG00000150672\"; gene_version \"18\"; ...      2172910  \n",
       "1601567     .  gene_id \"ENSG00000183117\"; gene_version \"20\"; ...      2059619  \n",
       "3030998     .  gene_id \"ENSG00000172264\"; gene_version \"18\"; ...      2057828  \n",
       "1104549     .  gene_id \"ENSG00000188107\"; gene_version \"15\"; ...      1987246  \n",
       "338522      .  gene_id \"ENSG00000168702\"; gene_version \"18\"; ...      1899593  \n",
       "2041854     .  gene_id \"ENSG00000183230\"; gene_version \"18\"; ...      1851180  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sorted.iloc[0:10,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "fb4f07a8-f229-47a4-a4e7-0360acd4fa2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_split = genes[\"group\"].str.split(\";\", expand = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "fc049bc3-a56c-432a-baea-4719c6d17d6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id \"ENSG00000160072\"</td>\n",
       "      <td>gene_version \"20\"</td>\n",
       "      <td>gene_name \"ATAD3B\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>gene_id \"ENSG00000234396\"</td>\n",
       "      <td>gene_version \"3\"</td>\n",
       "      <td>gene_source \"havana\"</td>\n",
       "      <td>gene_biotype \"lncRNA\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>gene_id \"ENSG00000225972\"</td>\n",
       "      <td>gene_version \"1\"</td>\n",
       "      <td>gene_name \"MTND1P23\"</td>\n",
       "      <td>gene_source \"havana\"</td>\n",
       "      <td>gene_biotype \"unprocessed_pseudogene\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>gene_id \"ENSG00000224315\"</td>\n",
       "      <td>gene_version \"2\"</td>\n",
       "      <td>gene_name \"RPL7P7\"</td>\n",
       "      <td>gene_source \"havana\"</td>\n",
       "      <td>gene_biotype \"processed_pseudogene\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>gene_id \"ENSG00000198744\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_name \"MTCO3P12\"</td>\n",
       "      <td>gene_source \"havana\"</td>\n",
       "      <td>gene_biotype \"unprocessed_pseudogene\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             0                   1                      2  \\\n",
       "0    gene_id \"ENSG00000160072\"   gene_version \"20\"     gene_name \"ATAD3B\"   \n",
       "107  gene_id \"ENSG00000234396\"    gene_version \"3\"   gene_source \"havana\"   \n",
       "111  gene_id \"ENSG00000225972\"    gene_version \"1\"   gene_name \"MTND1P23\"   \n",
       "114  gene_id \"ENSG00000224315\"    gene_version \"2\"     gene_name \"RPL7P7\"   \n",
       "117  gene_id \"ENSG00000198744\"    gene_version \"5\"   gene_name \"MTCO3P12\"   \n",
       "\n",
       "                                 3                                       4  \\\n",
       "0     gene_source \"ensembl_havana\"           gene_biotype \"protein_coding\"   \n",
       "107          gene_biotype \"lncRNA\"                                           \n",
       "111           gene_source \"havana\"   gene_biotype \"unprocessed_pseudogene\"   \n",
       "114           gene_source \"havana\"     gene_biotype \"processed_pseudogene\"   \n",
       "117           gene_source \"havana\"   gene_biotype \"unprocessed_pseudogene\"   \n",
       "\n",
       "        5  \n",
       "0          \n",
       "107  None  \n",
       "111        \n",
       "114        \n",
       "117        "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_split.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "2fb78064-9fb5-464d-8870-f5029c3e4e1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>107</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2212523</td>\n",
       "      <td>2220738</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234396\"; gene_version \"3\"; g...</td>\n",
       "      <td>8215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>111</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225972\"; gene_version \"1\"; g...</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>114</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8786211</td>\n",
       "      <td>8786913</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000224315\"; gene_version \"2\"; g...</td>\n",
       "      <td>702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>117</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>634376</td>\n",
       "      <td>634922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198744\"; gene_version \"5\"; g...</td>\n",
       "      <td>546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61855</th>\n",
       "      <td>3371150</td>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>4612</td>\n",
       "      <td>29626</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000271254\"; gene_version \"7\"; g...</td>\n",
       "      <td>25014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61856</th>\n",
       "      <td>3371224</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>30437</td>\n",
       "      <td>30580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275987\"; gene_version \"1\"; g...</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61857</th>\n",
       "      <td>3371227</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>35407</td>\n",
       "      <td>35916</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000268674\"; gene_version \"2\"; g...</td>\n",
       "      <td>509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61858</th>\n",
       "      <td>3371233</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>31698</td>\n",
       "      <td>32528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; g...</td>\n",
       "      <td>830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61859</th>\n",
       "      <td>3371241</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; g...</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61860 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         index     seqname          source feature    start      end score  \\\n",
       "0            0           1  ensembl_havana    gene  1471765  1497848     .   \n",
       "1          107           1          havana    gene  2212523  2220738     .   \n",
       "2          111           1          havana    gene   629062   629433     .   \n",
       "3          114           1          havana    gene  8786211  8786913     .   \n",
       "4          117           1          havana    gene   634376   634922     .   \n",
       "...        ...         ...             ...     ...      ...      ...   ...   \n",
       "61855  3371150  KI270711.1         ensembl    gene     4612    29626     .   \n",
       "61856  3371224  KI270713.1         ensembl    gene    30437    30580     .   \n",
       "61857  3371227  KI270713.1         ensembl    gene    35407    35916     .   \n",
       "61858  3371233  KI270713.1         ensembl    gene    31698    32528     .   \n",
       "61859  3371241  KI270713.1         ensembl    gene    21861    22024     .   \n",
       "\n",
       "      strand frame                                              group  \\\n",
       "0          +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...   \n",
       "1          +     .  gene_id \"ENSG00000234396\"; gene_version \"3\"; g...   \n",
       "2          +     .  gene_id \"ENSG00000225972\"; gene_version \"1\"; g...   \n",
       "3          -     .  gene_id \"ENSG00000224315\"; gene_version \"2\"; g...   \n",
       "4          +     .  gene_id \"ENSG00000198744\"; gene_version \"5\"; g...   \n",
       "...      ...   ...                                                ...   \n",
       "61855      -     .  gene_id \"ENSG00000271254\"; gene_version \"7\"; g...   \n",
       "61856      -     .  gene_id \"ENSG00000275987\"; gene_version \"1\"; g...   \n",
       "61857      +     .  gene_id \"ENSG00000268674\"; gene_version \"2\"; g...   \n",
       "61858      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; g...   \n",
       "61859      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; g...   \n",
       "\n",
       "       gene_length  \n",
       "0            26083  \n",
       "1             8215  \n",
       "2              371  \n",
       "3              702  \n",
       "4              546  \n",
       "...            ...  \n",
       "61855        25014  \n",
       "61856          143  \n",
       "61857          509  \n",
       "61858          830  \n",
       "61859          163  \n",
       "\n",
       "[61860 rows x 11 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "513f0ef1-6a1a-4aae-9f74-6d312a2ab4ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df = group_split.merge(genes, left_index = True, right_index = True,  how = \"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "8b0085ca-ab2f-4058-b4c3-118bdbc183f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id \"ENSG00000160072\"</td>\n",
       "      <td>gene_version \"20\"</td>\n",
       "      <td>gene_name \"ATAD3B\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>gene_id \"ENSG00000234396\"</td>\n",
       "      <td>gene_version \"3\"</td>\n",
       "      <td>gene_source \"havana\"</td>\n",
       "      <td>gene_biotype \"lncRNA\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2212523</td>\n",
       "      <td>2220738</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234396\"; gene_version \"3\"; g...</td>\n",
       "      <td>8215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>gene_id \"ENSG00000225972\"</td>\n",
       "      <td>gene_version \"1\"</td>\n",
       "      <td>gene_name \"MTND1P23\"</td>\n",
       "      <td>gene_source \"havana\"</td>\n",
       "      <td>gene_biotype \"unprocessed_pseudogene\"</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>629062</td>\n",
       "      <td>629433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225972\"; gene_version \"1\"; g...</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>gene_id \"ENSG00000224315\"</td>\n",
       "      <td>gene_version \"2\"</td>\n",
       "      <td>gene_name \"RPL7P7\"</td>\n",
       "      <td>gene_source \"havana\"</td>\n",
       "      <td>gene_biotype \"processed_pseudogene\"</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8786211</td>\n",
       "      <td>8786913</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000224315\"; gene_version \"2\"; g...</td>\n",
       "      <td>702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>gene_id \"ENSG00000198744\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_name \"MTCO3P12\"</td>\n",
       "      <td>gene_source \"havana\"</td>\n",
       "      <td>gene_biotype \"unprocessed_pseudogene\"</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>634376</td>\n",
       "      <td>634922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198744\"; gene_version \"5\"; g...</td>\n",
       "      <td>546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             0                   1                      2  \\\n",
       "0    gene_id \"ENSG00000160072\"   gene_version \"20\"     gene_name \"ATAD3B\"   \n",
       "107  gene_id \"ENSG00000234396\"    gene_version \"3\"   gene_source \"havana\"   \n",
       "111  gene_id \"ENSG00000225972\"    gene_version \"1\"   gene_name \"MTND1P23\"   \n",
       "114  gene_id \"ENSG00000224315\"    gene_version \"2\"     gene_name \"RPL7P7\"   \n",
       "117  gene_id \"ENSG00000198744\"    gene_version \"5\"   gene_name \"MTCO3P12\"   \n",
       "\n",
       "                                 3                                       4  \\\n",
       "0     gene_source \"ensembl_havana\"           gene_biotype \"protein_coding\"   \n",
       "107          gene_biotype \"lncRNA\"                                           \n",
       "111           gene_source \"havana\"   gene_biotype \"unprocessed_pseudogene\"   \n",
       "114           gene_source \"havana\"     gene_biotype \"processed_pseudogene\"   \n",
       "117           gene_source \"havana\"   gene_biotype \"unprocessed_pseudogene\"   \n",
       "\n",
       "        5 seqname          source feature    start      end score strand  \\\n",
       "0               1  ensembl_havana    gene  1471765  1497848     .      +   \n",
       "107  None       1          havana    gene  2212523  2220738     .      +   \n",
       "111             1          havana    gene   629062   629433     .      +   \n",
       "114             1          havana    gene  8786211  8786913     .      -   \n",
       "117             1          havana    gene   634376   634922     .      +   \n",
       "\n",
       "    frame                                              group  gene_length  \n",
       "0       .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "107     .  gene_id \"ENSG00000234396\"; gene_version \"3\"; g...         8215  \n",
       "111     .  gene_id \"ENSG00000225972\"; gene_version \"1\"; g...          371  \n",
       "114     .  gene_id \"ENSG00000224315\"; gene_version \"2\"; g...          702  \n",
       "117     .  gene_id \"ENSG00000198744\"; gene_version \"5\"; g...          546  "
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "90f7b802-3776-416f-81aa-593847a7b9cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_1 = genes.merge(ens_df, on = [\"seqname\", \"end\", \"start\"], how = \"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "a000d7a4-7889-449b-834c-7047f77c6b5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source_x</th>\n",
       "      <th>feature_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score_x</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>frame_x</th>\n",
       "      <th>group_x</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>source_y</th>\n",
       "      <th>feature_y</th>\n",
       "      <th>score_y</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>frame_y</th>\n",
       "      <th>group_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "      <td>369453</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>9833</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "      <td>81894</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname        source_x feature_x    start      end score_x strand_x  \\\n",
       "0       1  ensembl_havana      gene  1471765  1497848       .        +   \n",
       "1       1  ensembl_havana      gene  1471765  1497848       .        +   \n",
       "2       1  ensembl_havana      gene  3069168  3438621       .        +   \n",
       "3       1  ensembl_havana      gene  2403964  2413797       .        -   \n",
       "4       1  ensembl_havana      gene  2228319  2310213       .        +   \n",
       "\n",
       "  frame_x                                            group_x  gene_length  \\\n",
       "0       .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083   \n",
       "1       .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083   \n",
       "2       .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...       369453   \n",
       "3       .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...         9833   \n",
       "4       .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...        81894   \n",
       "\n",
       "         source_y   feature_y score_y strand_y frame_y  \\\n",
       "0  ensembl_havana        gene       .        +       .   \n",
       "1  ensembl_havana  transcript       .        +       .   \n",
       "2  ensembl_havana        gene       .        +       .   \n",
       "3  ensembl_havana        gene       .        -       .   \n",
       "4  ensembl_havana        gene       .        +       .   \n",
       "\n",
       "                                             group_y  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  \n",
       "3  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  \n",
       "4  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d228ba88-bbc1-4a25-b324-98f0da4017f5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
